src/megaapi_impl.cpp

```diff
@@ -23829,14 +23890,73 @@ void MegaFolderUploadController::cancel()
     //remove subtransfers from pending transferQueue
     megaApi->cancelPendingTransfersByFolderTag(tag);
 
+    //remove ongoing subtransfers
+    long long cancelledSubTransfers = 0;
+    DBTableTransactionCommitter committer(client->tctable);
     while (!subTransfers.empty())
     {
         auto subTransfer = *subTransfers.begin();
-        subTransfer->setState(MegaTransfer::STATE_COMPLETED);
-        DBTableTransactionCommitter committer(client->tctable);
-        megaApi->fireOnTransferFinish(subTransfer, make_unique<MegaErrorPrivate>(API_EINCOMPLETE), committer);
+
+        Transfer *transfer = subTransfer->getTransfer();
+        if (!transfer)
+        {
+            LOG_warn << "Subtransfer without attached Transfer for folder transfer: " << subTransfer->getFileName();
+
+            subTransfer->setState(MegaTransfer::STATE_COMPLETED); //TODO: copmleted?
+            megaApi->fireOnTransferFinish(subTransfer, make_unique<MegaErrorPrivate>(API_EINCOMPLETE), committer);
+
+            continue;
+        }
+
+#ifdef _WIN32
+        if (transfer->type == GET)
+        {
+            transfer->localfilename.append("", 1);
+            WIN32_FILE_ATTRIBUTE_DATA fad;
+            if (GetFileAttributesExW((LPCWSTR)transfer->localfilename.data(), GetFileExInfoStandard, &fad))
+                SetFileAttributesW((LPCWSTR)transfer->localfilename.data(), fad.dwFileAttributes & ~FILE_ATTRIBUTE_HIDDEN);
+            transfer->localfilename.resize(transfer->localfilename.size()-1);
+        }
+#endif
+
+        MegaErrorPrivate megaError(API_EINCOMPLETE);
+        subTransfer->setLastError(&megaError);
+
+        bool found = false;
+        file_list files = transfer->files;
+        file_list::iterator iterator = files.begin();
+        while (iterator != files.end())
+        {
+            File *file = *iterator;
+            iterator++;
+            if (file->tag == subTransfer->getTag())
```

Comment:
I think every file download that is part of the folder download gets the same tag?   So we won't necessarily match the exact transfer file for this MegaTransfer - perhaps we should remove all that have the matching tag, without breaking the loop?

Reply:
I too borrowed the code from TYPE_CANCEL_TRANSFER. 
> I think every file download that is part of the folder download gets the same tag? 
they get the same getFolderTransferTag, not the same tag, don't they?

I'm concerned that there could be more File* matching the MegaTransfer::tag. that would be happening in TYPE_CANCEL_TRANSFER too :thinking: . and breaking the loop could be incorrect ...

Reply:
Second read: no concerns: only one File will match MegaTransfer::tag. File::tag is unique AFAIK

Reply:
Ah ok, good to know, thanks for checking that :-)